packages/@lwc/node-reactions/src/traverse.ts

```diff
@@ -31,12 +35,15 @@ export default function queueReactionsForTree(
         });
     }
 
-    if (root instanceof Element && !!root.shadowRoot) {
+    const nodeType = nodeTypeGetter.call(root);
+    // Perf optimization: Only Element type can have shadowRoot
+    if (nodeType === 1 && (root as Element).shadowRoot) {
```

Comment:
I'm worry about this check here. if the element has a closed shadowRoot, we can't do much. That's why I think my suggestion was to patch `attachShadow()` so we can always grab the shadowRoot when needed, and we can have a custom mechanism to determine shadow. Something like:

```js
if (!isUndefined(root.$$myInternalFn$$)) {
     const sr = root.$$myInternalFn$$();
}
```

Reply:
the patch on attachShadow will create `$$myInternalFn$$` or a symbol to provide a fn to access the shadow root instance.

Reply:
Yup, agreed. Closed shadow was an open question in the RFC too.
About patching attachShadow and adding a field, if I put my security hat on: What stops external users from utilizing the same field to bypass closed shadow?